File: src/java/principles_systems_architecture/chapter2/section1/State.java
package principles_systems_architecture.chapter2.section1;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

enum State {
  審査中,
  承認済,
  実施中,
  終了,
  差し戻し中,
  中断中
}

class StateTransitions {
  Map<State, Set<State>> allowed;

  {
    allowed = new HashMap<>();
    allowed.put(State.審査中, EnumSet.of(State.承認済, State.差し戻し中));
    allowed.put(State.差し戻し中, EnumSet.of(State.審査中, State.終了));
    allowed.put(State.承認済, EnumSet.of(State.実施中, State.終了));
    allowed.put(State.実施中, EnumSet.of(State.中断中, State.終了));
    allowed.put(State.中断中, EnumSet.of(State.実施中, State.終了));
  }

  boolean canTransit(State from, State to) {
    return allowed.get(from).contains(to);
  }
}

File: src/clj/principles_systems_architecture/chapter2/section1/state.clj
(ns principles-systems-architecture.chapter2.section1.state
  (:require [schema.core :as s]))

(def State
  (s/enum :審査中
          :承認済
          :実施中
          :終了
          :差し戻し中
          :中断中))

(s/def allowed :- {State #{State}}
  {:審査中 #{:承認済 :差し戻し中}
   :差し戻し中 #{:審査中 :終了}
   :承認済 #{:実施中 :終了}
   :実施中 #{:中断中 :終了}
   :中断中 #{:実施中 :終了}})

(s/defn can-transit? :- s/Bool
  [from :- State
   to :- State]
  (contains? (get allowed from) to))
